File: DataStructures/Tries/trie.js
/****************************************************************/
/*Tries  ********/
/****************************************************************/
class Node {
  constructor(){
    this.keys = new Map();
    this.endOfWord = false;
  }
}
class Trie {
  constructor() {
    this.root = new Node();
  }
  buildList(node=this.root, string=new String(), arr=new Array()){
    for(let charac of node.keys.keys()){
      this.buildList(node.keys.get(charac), string.concat(charac), arr);
    }
    if(node.endOfWord){ 
      arr.push(string); 
    }

    return arr;
  }
  searchPrefix(prefix){
    /*  
      - Descr.: Print out all the words containing a specific prefix.
      - Input:  It receives the prefix it must look for as input.
      - Output: It stores all the words which begin with that prefix and print them out.
    */
    let prefixMatch = new Array();
    let allWords = this.buildList();

    if(!allWords.length){ console.log('Trie is empty'); return;}

    allWords.forEach(element => {
      if(element.includes(prefix)){ prefixMatch.push(element); }
    });

    if(prefixMatch.length){ 
      console.log(prefixMatch); return prefixMatch; 
    }
    console.log(`No match for the prefix/word "${prefix}".`);
  }
  insertWord(str, node = this.root){
    /*
      - Descr.: Inserts a new word in the trie.
      - Input:  It receives the word to be inserted as arguments and cheks for repeated keys. 
      - Output: Print out the words on the trie.
    */
    if(!str.length) { node.endOfWord = true; return; }
    if(!node.keys.has(str[0])){
      node.keys.set(str[0], new Node());
    }
    this.insertWord(str.slice(1), node.keys.get(str[0]));
  }
  removeWord(word, node = this.root, index = 0){
    /*
      - Descr.: Remove a word from the trie.
      - Input:  It receives the word to be removed as argument and checks if this word is on the tree. 
      - Output: Removes the word completely or partially depending on whether part of the word is a prefix to  other words on the trie. Lastly, print out the words on the trie.
        When deleting a string form a Trie, we can consider the cases below:
    */
    let currentCharac = word[index];
    let currentNode = node.keys.get(currentCharac);
    let canWeDeleteThisNode = false;

    //-- CASE#1: The word being removed has the same prefix as some other word in the Trie. In this case, we cannot delete the entire string, but only the independent "members". Ex: mind and mid.
    if(this.hasDependent(currentNode)){
      this.removeWord(word, currentNode, index+1);
      return false;
    }

    //-- CASE#2: The entire word being removed is a prefix of some other word. In this case, we don't delete anything and just set the flag indicating the end of the word being removed as FALSE. Ex.: mid and middle.
    if(index === word.length - 1){
      if(currentNode.keys.size){
        currentNode.endOfWord = false;
        return false;
      }
      else{ 
        node.keys.delete(currentCharac);
        return true;
      }
    }

    //-- CASE#3: Some other word is a prefix of the word we want to remove. As with CASE#1, we only delete the independent "members". Ex.: middle and mid.
    if(currentNode.endOfWord){
      this.removeWord(word, currentNode, index+1);
      return false;
    }

    //-- CASE#4: No one is dependent on the word we want to remove, we just delete it. Ex.: apple.
    canWeDeleteThisNode = this.removeWord(word, currentNode, index+1);

    if(canWeDeleteThisNode){
      node.keys.delete(currentCharac);
    }
  }
  hasDependent(node){
    return ((node.keys.size > 1) ? true : false);
  }
  isOnTree(str){
    /*
      - Descr.: Checks if a word is on the Trie.
      - Input:  It receives the word to be searched as argument. 
      - Output: Returns a boolean indicating whether the word is present or not.
    */
    return ((this.buildList().includes(str)) ? true : false);
  }
  printTrie(){
    /*
      - Descr.: Prints out all the words on the trie (Try to print in alphabetical order later).
      - Input:  - 
      - Output: All the words on the trie are logged on the screen.
    */
    let allWords = this.buildList();

    if(allWords.length){ 
      console.log(allWords); return; 
    }
    console.log('Trie is empty');
  }
}
const myTrie = new Trie();
myTrie.insertWord('man');
myTrie.insertWord('mind');
myTrie.insertWord('mid');
myTrie.insertWord('apple');
myTrie.insertWord('tree');
myTrie.insertWord('town');
myTrie.insertWord('middle');
myTrie.insertWord('app');

myTrie.printTrie();

myTrie.searchPrefix('mi');
myTrie.searchPrefix('app');
myTrie.searchPrefix('au');

console.log(myTrie.isOnTree('man'));
console.log(myTrie.isOnTree('tow'));

myTrie.removeWord('apple');
console.log(myTrie.isOnTree('apple'));
myTrie.printTrie();

console.log(myTrie);
/*
          root
      -------------
      /     |     \
    m       a      t
  /  \      |    /  \
a     i     p   r    o
|   /  \   |    |     | 
n* n   d*  p    e     w
  |    |   |    |     |
  d*   d   e*   e*   n*
      |
      l
      |
      e*

*/

File: DataStructures/Tries/trie.cpp
/***
Trie -- C++
*/

#include <iostream>
#include <vector>

class TrieNode {
public:
  // array of pointers to 'TreeNode' (26: letters on the English alphabet):
  TrieNode *children[26];
  // boolean used to indicate that the current TreeNode corresponds to the end of a word:
  bool endOfWord;
  // counter which indicates how many words the current TreeNode is part of:
  int count;

  // constructor:
  TrieNode() {
    endOfWord = false;
    count = 0;

    // initialize all pointers to null:
    for (size_t i{0}; i < 26; i++) {
      children[i] = nullptr;
    }
  }
};

class Trie {
private:
  TrieNode *root;

public:
  void insertWord(std::string word);
  void deleteWord(std::string word);
  bool searchWord(std::string word);

  bool deleteHelper(std::string word, TrieNode *parent, int idx);
  bool hasDependent(TrieNode *node);
  int searchPrefix(std::string word);

  Trie() {
    this->root = new TrieNode();
  }
  ~Trie() {}
};

void Trie::insertWord(std::string word) {
  TrieNode *current = this->root;

  for (auto i{0}; i < word.size(); i++) {
    // calculate the position in the 'children' array by subtracting the
    // ascii code for the character word[i] by the ascii code for 'a':
    auto j = int(word.at(i)) - 97; // ascii('a') = 97

    // check if there is already a word which uses the letter word[i],
    // if not, initialize a new node at that position:
    if (current->children[j] == nullptr) {
      current->children[j] = new TrieNode();
    }

    // increment the word count for node children[j] since a new word is being
    // added which this node is a part of:
    ++current->children[j]->count;

    // advance to the next character in the word:
    current = current->children[j];
  }
  // set the flag:
  current->endOfWord = true;
}

void Trie::deleteWord(std::string word) {
  // check if 'word' exists in the trie before trying to delete it:
  if (searchWord(word)) {
    deleteHelper(word, this->root, 0);
  }
}

bool Trie::searchWord(std::string word) {
  TrieNode *current = this->root;

  for (auto i{0}; i < word.size(); i++) {
    auto j = int(word.at(i)) - 97;

    if (current->children[j] == nullptr) {
      return false;
    }

    current = current->children[j];
  }

  return (current->endOfWord);
}

bool Trie::deleteHelper(std::string word, TrieNode *parent, int idx) {
  char currentCharac = word.at(idx);
  TrieNode *current = parent->children[int(currentCharac) - 97];
  bool removable = false;

  // case #1 - current character is a prefix and we have not reached the end of the word:
  if (current->count && idx + 1 < word.size()) {
    // recurse down the trie until we've traversed the whole 'word'.
    deleteHelper(word, current, idx + 1);
    // decrease word count for the current character:
    parent->children[int(currentCharac) - 97]->count--;
    // return false because the nodes itself cannot be deleted:
    return false;
  }
  // case #2 - we have reached the end of the word:
  if (idx == word.size() - 1) {
    // case #2.a - current character is a prefix, so the node itself cannot be deleted:
    if (current->count) {
      // decrease word count for the current character:
      parent->children[int(currentCharac) - 97]->count--;
      // set the 'endOfWord' flag to false since the word ending at this character is being deleted:
      current->endOfWord = false;
      // return false because the nodes itself cannot be deleted:
      return false;
    }
    // case #2.b - current character is not a prefix, so the node can be safely deleted:
    else {
      parent->children[int(currentCharac) - 97]->count--;
      current = nullptr;

      return true;
    }
  }

  if (current->endOfWord && idx + 1 < word.size()) {
    deleteHelper(word, current, idx + 1);
    return false;
  }

  removable = deleteHelper(word, current, idx + 1);

  if (removable) {
    parent->children[int(currentCharac) - 97]->count--;
    current = nullptr;
  }

  return false;
}

int Trie::searchPrefix(std::string word) {
  TrieNode *current = this->root;

  for (size_t i{0}; i < word.length(); i++) {
    size_t j = int(word.at(i)) - 97;
    if (current->children[j] == nullptr) {
      return 0;
    }
    current = current->children[j];
  }
  return current->count;
}

int main() {
  std::cout.setf(std::ios::boolalpha);

  Trie myTrie;

  myTrie.insertWord("trial");
  myTrie.insertWord("tree");
  myTrie.insertWord("man");
  myTrie.insertWord("mind");
  myTrie.insertWord("mid");
  myTrie.insertWord("apple");
  myTrie.insertWord("middle");

  std::cout << "'trial' is on Trie: " << myTrie.searchWord("trial") << "\n";
  std::cout << "'time' is on Trie: " << myTrie.searchWord("time") << "\n";

  std::cout << "prefix 'tr': " << myTrie.searchPrefix("tr") << "\n";
  std::cout << "prefix 'mi': " << myTrie.searchPrefix("mi") << "\n";

  myTrie.deleteWord("trial");
  std::cout << "'trial' is on Trie: " << myTrie.searchWord("trial") << "\n";

  myTrie.deleteWord("mind");
  std::cout << "'mind' is on Trie: " << myTrie.searchWord("mind") << "\n";
  std::cout << "prefix 'mi': " << myTrie.searchPrefix("mi") << "\n";

  return 0;
}